In [1]:
import pandas as pd
import numpy as np

In [2]:
cols = ['bookID' ,'title',  'authors', 'average_rating', 'isbn', 'isbn13', 'language_code', 'num_pages', 'ratings_count',      
        'text_reviews_count', 'publication_date', 'publisher',  'coauthor']
df = pd.read_csv('books_data.csv', names=cols, skiprows=1,engine='python', quotechar=',', on_bad_lines='warn')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   bookID              11127 non-null  int64 
 1   title               11127 non-null  object
 2   authors             11127 non-null  object
 3   average_rating      11127 non-null  object
 4   isbn                11127 non-null  object
 5   isbn13              11127 non-null  object
 6   language_code       11127 non-null  object
 7   num_pages           11127 non-null  object
 8   ratings_count       11127 non-null  int64 
 9   text_reviews_count  11127 non-null  int64 
 10  publication_date    11127 non-null  object
 11  publisher           11127 non-null  object
 12  coauthor            4 non-null      object
dtypes: int64(3), object(10)
memory usage: 1.1+ MB


In [4]:
df.head(2)

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   

                      authors average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré           4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré           4.49  0439358078  9780439358071   

  language_code num_pages  ratings_count  text_reviews_count publication_date  \
0           eng       652        2095690               27591        9/16/2006   
1           eng       870        2153167               29221         9/1/2004   

         publisher coauthor  
0  Scholastic Inc.     None  
1  Scholastic Inc.     None

In [5]:
dirty = df.loc[~df['coauthor'].isnull(), : ]

In [6]:
dirty

bookID                                              title  \
3348   12224  Streetcar Suburbs: The Process of Growth in Bo...   
4702   16914                  The Tolkien Fan's Medieval Reader   
5877   22128                     Patriots (The Coming Collapse)   
8979   34889  Brown's Star Atlas: Showing All The Bright Sta...   

                                       authors  \
3348                           Sam Bass Warner   
4702  David E. Smith (Turgon of TheOneRing.net   
5877                              James Wesley   
8979                                     Brown   

                                         average_rating  isbn      isbn13  \
3348                                  Jr./Sam B. Warner  3.58  0674842111   
4702   one of the founding members of this Tolkien w...  3.58  1593600119   
5877                                             Rawles  3.63  156384155X   
8979                                     Son & Ferguson  0.00  0851742718   

      language_code num_pages  ratings_count  text_reviews_count  \
3348  9780674842113     en-US            236                  61   
4702  9781593600112       eng            400                  26   
5877  9781563841552       eng            342                  38   
8979  9780851742717       eng             49                   0   

     publication_date  publisher                     coauthor  
3348                6  4/20/2004     Harvard University Press  
4702                4   4/6/2004          Cold Spring Press\t  
5877                4  1/15/1999  Huntington House Publishers  
8979                0   5/1/1977    Brown Son & Ferguson Ltd.

In [7]:
coauthor = dirty.loc[:, 'average_rating']
coauthor = coauthor.rename("author2")

In [8]:
right = dirty.loc[:, 'average_rating':'coauthor'].shift(periods=-1, axis=1)

In [9]:
left = dirty.loc[:, :'authors']

In [10]:
left['authors'] = left['authors'] + coauthor

In [11]:
left

bookID                                              title  \
3348   12224  Streetcar Suburbs: The Process of Growth in Bo...   
4702   16914                  The Tolkien Fan's Medieval Reader   
5877   22128                     Patriots (The Coming Collapse)   
8979   34889  Brown's Star Atlas: Showing All The Bright Sta...   

                                                authors  
3348                  Sam Bass Warner Jr./Sam B. Warner  
4702  David E. Smith (Turgon of TheOneRing.net one o...  
5877                                James Wesley Rawles  
8979                               Brown Son & Ferguson

In [12]:
right

average_rating        isbn         isbn13 language_code  num_pages  \
3348           3.58  0674842111  9780674842113         en-US        236   
4702           3.58  1593600119  9781593600112           eng        400   
5877           3.63  156384155X  9781563841552           eng        342   
8979           0.00  0851742718  9780851742717           eng         49   

      ratings_count text_reviews_count publication_date  \
3348             61                  6        4/20/2004   
4702             26                  4         4/6/2004   
5877             38                  4        1/15/1999   
8979              0                  0         5/1/1977   

                        publisher coauthor  
3348     Harvard University Press     None  
4702          Cold Spring Press\t     None  
5877  Huntington House Publishers     None  
8979    Brown Son & Ferguson Ltd.     None

In [13]:
clean = pd.concat([left,  right, coauthor], axis=1)

In [14]:
clean = clean.drop(columns= ['coauthor'])

In [15]:
clean

bookID                                              title  \
3348   12224  Streetcar Suburbs: The Process of Growth in Bo...   
4702   16914                  The Tolkien Fan's Medieval Reader   
5877   22128                     Patriots (The Coming Collapse)   
8979   34889  Brown's Star Atlas: Showing All The Bright Sta...   

                                                authors average_rating  \
3348                  Sam Bass Warner Jr./Sam B. Warner           3.58   
4702  David E. Smith (Turgon of TheOneRing.net one o...           3.58   
5877                                James Wesley Rawles           3.63   
8979                               Brown Son & Ferguson           0.00   

            isbn         isbn13 language_code  num_pages  ratings_count  \
3348  0674842111  9780674842113         en-US        236             61   
4702  1593600119  9781593600112           eng        400             26   
5877  156384155X  9781563841552           eng        342             38   
8979  0851742718  9780851742717           eng         49              0   

     text_reviews_count publication_date                    publisher  \
3348                  6        4/20/2004     Harvard University Press   
4702                  4         4/6/2004          Cold Spring Press\t   
5877                  4        1/15/1999  Huntington House Publishers   
8979                  0         5/1/1977    Brown Son & Ferguson Ltd.   

                                                author2  
3348                                  Jr./Sam B. Warner  
4702   one of the founding members of this Tolkien w...  
5877                                             Rawles  
8979                                     Son & Ferguson

In [16]:
df.loc[clean.index, :] = clean

In [17]:
df = df.drop(columns= ['coauthor', 'bookID', 'isbn', 'isbn13'])

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   title               11127 non-null  object
 1   authors             11127 non-null  object
 2   average_rating      11127 non-null  object
 3   language_code       11127 non-null  object
 4   num_pages           11127 non-null  object
 5   ratings_count       11127 non-null  int64 
 6   text_reviews_count  11127 non-null  object
 7   publication_date    11127 non-null  object
 8   publisher           11127 non-null  object
dtypes: int64(1), object(8)
memory usage: 782.5+ KB


In [19]:
df['average_rating'] = df['average_rating'].astype(np.float32)
df['num_pages'] = df['num_pages'].astype(np.int32)
df['ratings_count'] = df['ratings_count'].astype(np.int32)
df['text_reviews_count'] = df['text_reviews_count'].astype(np.int32)

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   title               11127 non-null  object 
 1   authors             11127 non-null  object 
 2   average_rating      11127 non-null  float32
 3   language_code       11127 non-null  object 
 4   num_pages           11127 non-null  int32  
 5   ratings_count       11127 non-null  int32  
 6   text_reviews_count  11127 non-null  int32  
 7   publication_date    11127 non-null  object 
 8   publisher           11127 non-null  object 
dtypes: float32(1), int32(3), object(5)
memory usage: 608.6+ KB


In [21]:
df.iloc[3348, :]

title                 Streetcar Suburbs: The Process of Growth in Bo...
authors                               Sam Bass Warner Jr./Sam B. Warner
average_rating                                                     3.58
language_code                                                     en-US
num_pages                                                           236
ratings_count                                                        61
text_reviews_count                                                    6
publication_date                                              4/20/2004
publisher                                      Harvard University Press
Name: 3348, dtype: object

In [22]:
dates = pd.to_datetime(df.publication_date, format="%m/%d/%Y", errors="coerce")

In [23]:
dates[dates.isnull()]

8180    NaT
11098   NaT
Name: publication_date, dtype: datetime64[ns]

In [24]:
df.publication_date[8180]

'11/31/2000'

In [25]:
df.publication_date[11098]

'6/31/1982'

In [26]:
df.iloc[8180, 7] = '11/30/2000' 
df.iloc[11098, 7] = '6/30/1982'

df.publication_date = pd.to_datetime(df.publication_date, format="%m/%d/%Y", errors="raise")

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   title               11127 non-null  object        
 1   authors             11127 non-null  object        
 2   average_rating      11127 non-null  float32       
 3   language_code       11127 non-null  object        
 4   num_pages           11127 non-null  int32         
 5   ratings_count       11127 non-null  int32         
 6   text_reviews_count  11127 non-null  int32         
 7   publication_date    11127 non-null  datetime64[ns]
 8   publisher           11127 non-null  object        
dtypes: datetime64[ns](1), float32(1), int32(3), object(4)
memory usage: 608.6+ KB


In [28]:
df.publication_date = (df.publication_date - pd.to_datetime('1970-01-01', format='%Y-%m-%d')).dt.days

In [29]:
author_name=[]
for i in range(1, df["authors"].str.split("/", expand=True).shape[1]):
    author_name.append(f"Author_{i}")
df  = pd.concat((df, df["authors"].str.split("/", expand=True)), axis=1)

In [30]:
 df = df.drop(columns= ['authors'])

In [31]:
df.head(2)

title  average_rating  \
0  Harry Potter and the Half-Blood Prince (Harry ...            4.57   
1  Harry Potter and the Order of the Phoenix (Har...            4.49   

  language_code  num_pages  ratings_count  text_reviews_count  \
0           eng        652        2095690               27591   
1           eng        870        2153167               29221   

   publication_date        publisher             0              1  ...    41  \
0             13407  Scholastic Inc.  J.K. Rowling  Mary GrandPré  ...  None   
1             12662  Scholastic Inc.  J.K. Rowling  Mary GrandPré  ...  None   

     42    43    44    45    46    47    48    49    50  
0  None  None  None  None  None  None  None  None  None  
1  None  None  None  None  None  None  None  None  None  

[2 rows x 59 columns]

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11127 entries, 0 to 11126
Data columns (total 59 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   title               11127 non-null  object 
 1   average_rating      11127 non-null  float32
 2   language_code       11127 non-null  object 
 3   num_pages           11127 non-null  int32  
 4   ratings_count       11127 non-null  int32  
 5   text_reviews_count  11127 non-null  int32  
 6   publication_date    11127 non-null  int64  
 7   publisher           11127 non-null  object 
 8   0                   11127 non-null  object 
 9   1                   4564 non-null   object 
 10  2                   1485 non-null   object 
 11  3                   479 non-null    object 
 12  4                   257 non-null    object 
 13  5                   192 non-null    object 
 14  6                   138 non-null    object 
 15  7                   117 non-null    object 
 16  8   

 0   bookID              11127 non-null  int64  
 1   title               11127 non-null  object
 2   authors             11127 non-null  object
 3   average_rating      11127 non-null  object
 4   isbn                11127 non-null  object
 5   isbn13              11127 non-null  object
 6   language_code       11127 non-null  object
 7     num_pages         11127 non-null  object
 8   ratings_count       11127 non-null  int64 
 9   text_reviews_count  11127 non-null  int64 
 10  publication_date    11127 non-null  object
 11  publisher           11127 non-null  object
 12  Unnamed: 12         4 non-null      object
        
        
1. First drop the four rows that have the 12 columns.  I will write the code to shift the data to left for these rows. For now you can simply drop the rows. 
Syntax
df.loc[~df['Unnamed: 12'].isnull(), : ]

2. Convert num pages, avergae rating to numbers.
3. Drop bookID, isbn,  isbn13, Unnamed: 12 
4. Convert date time to time delta from 1 Jan 1970. Meaning convert  publication_date to a date and subtract  1 Jan 1970 from it. 
5. Split the co authors to an another column. Difficult I will do. 

After the data cleaning step we will have

title, average_rating, language_code, ratings_count, text_reviews_count, publication_date, publisher, author1 , author2, to max authors etc.